src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.GeneratedFileReplacingCompilationTracker.cs

```diff
@@ -0,0 +1,190 @@
+        {
+            private readonly ICompilationTracker _underlyingTracker;
+            private readonly SourceGeneratedDocumentState _replacedGeneratedDocumentState;
+
+            /// <summary>
+            /// The lazily-produced compilation that has the generated document updated. This is initialized by call to
+            /// <see cref="EnsureNewCompilationComputedAsync"/>, and since it's only written once is safe to read if non-null.
+            /// </summary>
+            [DisallowNull]
+            private Compilation? _compilationWithReplacement;
+
+            public GeneratedFileReplacingCompilationTracker(ICompilationTracker underlyingTracker, SourceGeneratedDocumentState replacementDocumentState)
+            {
+                _underlyingTracker = underlyingTracker;
+                _replacedGeneratedDocumentState = replacementDocumentState;
+            }
+
+            [MemberNotNull(nameof(_compilationWithReplacement))]
+            private async Task EnsureNewCompilationComputedAsync(SolutionState solutionState, CancellationToken cancellationToken)
+            {
+#pragma warning disable CS8774 // Member must have a non-null value when exiting -- this is broken with await
+
+                // Fast path if we've definitely already done this before
+                if (_compilationWithReplacement != null)
+                {
+                    return;
+                }
+
+                var underlyingCompilation = await _underlyingTracker.GetCompilationAsync(solutionState, cancellationToken).ConfigureAwait(false);
+                var underlyingSourceGeneratedDocuments = await _underlyingTracker.GetSourceGeneratedDocumentStatesAsync(solutionState, cancellationToken).ConfigureAwait(false);
+
+                underlyingSourceGeneratedDocuments.TryGetState(_replacedGeneratedDocumentState.Id, out var existingState);
+
+                Compilation newCompilation;
+
+                var newSyntaxTree = await _replacedGeneratedDocumentState.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+
+                if (existingState != null)
+                {
+                    var existingSyntaxTree = await existingState.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+                    newCompilation = underlyingCompilation.ReplaceSyntaxTree(existingSyntaxTree, newSyntaxTree);
+                }
+                else
+                {
+
+                    // The existing output no longer exists in the underlying compilation. This could happen if the user made
+                    // an edit which would cause this file to no longer exist, but they're still operating on an open representation
+                    // of that file. To ensure that this snapshot is still usable, we'll just add this document back in. This is not a
+                    // semantically correct operation, but working on stale snapshots never has that guarantee.
```

Comment:
considering doc'ing the true part of the if statement.  thanks for this comment, very helpful.